File: mobile/frontend/lib/data/repositories/user_repository.dart
import 'package:frontend/data/models/user.dart';
import 'package:frontend/data/enums.dart';
import 'base_repository.dart';

class UserRepository extends BaseRepository<User> {
  UserRepository() : super('users');

  @override
  User fromMap(Map<String, dynamic> map) => User.fromMap(map);

  @override
  Map<String, dynamic> toMap(User entity) => entity.toMap();

  // ── Domain queries ────────────────────────────────────────

  /// Find user by email (for login).
  Future<User?> findByEmail(String email) => findOne('email', email);

  /// Authenticate by email + password. Returns null if not found.
  Future<User?> authenticate(String email, String password) async {
    final results = await query(
      where: 'email = ? AND password = ? AND is_active = 1',
      whereArgs: [email, password],
      limit: 1,
    );
    return results.isEmpty ? null : results.first;
  }

  /// Get users by role.
  Future<List<User>> getByRole(UserRole role) {
    return query(where: 'role = ?', whereArgs: [role.value]);
  }

  /// Get active users only.
  Future<List<User>> getActiveUsers() {
    return query(where: 'is_active = 1', orderBy: 'name ASC');
  }

  /// Get all users sorted by name.
  Future<List<User>> getAllSorted() {
    return getAll(orderBy: 'name ASC');
  }

  /// Toggle user active status.
  Future<void> toggleActive(String userId) async {
    final user = await getById(userId);
    if (user != null) {
      await update(userId, {'is_active': user.isActive ? 0 : 1});
    }
  }
}

File: mobile/backend/app/repositories/user_repository.py
"""
User repository for Firestore operations.
"""

from typing import List, Optional, Dict, Any

from app.repositories.base_repository import BaseRepository
from app.core.enums import UserRole


class UserRepository(BaseRepository):
    """Repository for User documents."""

    def __init__(self):
        super().__init__("users")

    async def get_by_email(self, email: str) -> Optional[Dict[str, Any]]:
        """Find a user by email address."""
        return await self.find_one("email", email)

    async def get_by_role(self, role: UserRole) -> List[Dict[str, Any]]:
        """Get all users with a specific role."""
        return await self.query(filters=[("role", "==", role.value)])

    async def get_employees(self) -> List[Dict[str, Any]]:
        """Get all employees."""
        return await self.query(filters=[("role", "==", UserRole.EMPLOYEE.value)])

    async def get_active_employees(self) -> List[Dict[str, Any]]:
        """Get all active employees (is_active=True)."""
        employees = await self.get_employees()
        return [e for e in employees if e.get("is_active", True)]

    async def get_filtered(
        self, role: Optional[UserRole] = None
    ) -> List[Dict[str, Any]]:
        """Get users with optional role filter."""
        filters = []
        if role is not None:
            filters.append(("role", "==", role.value))
        return await self.query(filters=filters if filters else None)
